drivers/firmware/ti_sci/ti_sci.c

```diff
@@ -0,0 +1,1660 @@
+}
+
+static bool ti_sci_is_response_ack(void *r)
+{
+	struct ti_sci_msg_hdr *hdr = (struct ti_sci_msg_hdr *)r;
+
+	return hdr->flags & TI_SCI_FLAG_RESP_GENERIC_ACK ? true : false;
+}
+
+static int ti_sci_get_response(const struct device *dev, struct ti_sci_xfer *xfer)
+{
+	struct tisci_data *dev_data = dev->data;
+	struct ti_sci_msg_hdr *hdr;
+	int ret = 0;
+
+	if (!xfer->rx_message.buf) {
+		LOG_ERR("No response buffer provided");
+		return -EINVAL;
+	}
+
+	if (k_sem_take(&response_ready_sem, K_MSEC(dev_data->desc.max_rx_timeout_ms * 1000)) != 0) {
+		LOG_ERR("Timeout waiting for response");
+		return -ETIMEDOUT;
+	}
+
+	hdr = (struct ti_sci_msg_hdr *)rx_message.buf;
+
+	/* Sanity check for message response */
+	if (hdr->seq != dev_data->seq) {
+		LOG_ERR("HDR seq != data seq [%d != %d]\n", hdr->seq, dev_data->seq);
+		return -EINVAL;
+	}
+
+	if (rx_message.size > dev_data->desc.max_msg_size) {
+		LOG_ERR("rx_message.size [ %d ] > max_msg_size\n", xfer->rx_message.size);
+		return -EINVAL;
+	}
+
+	if (rx_message.size < xfer->rx_message.size) {
+		LOG_ERR("rx_message.size [ %d ] < xfer->rx_message.size\n", xfer->rx_message.size);
+		return -EINVAL;
+	}
+
+	memcpy(xfer->rx_message.buf, rx_message.buf, xfer->rx_message.size);
+
+	return ret;
+}
+
+static int ti_sci_do_xfer(const struct device *dev, struct ti_sci_xfer *xfer)
+{
+	const struct tisci_config *config = dev->config;
+	struct mbox_msg *msg = &xfer->tx_message;
+	int ret;
+
+	ret = mbox_send_dt(&config->mbox_tx, msg);
+	if (ret < 0) {
+		LOG_ERR("Could not send (%d)\n", ret);
+		return ret;
+	}
+
+	/* Get response if requested */
+	if (xfer->rx_message.size) {
+		return ti_sci_get_response(dev, xfer);
+	}
+
+	if (!ti_sci_is_response_ack(rx_message.buf)) {
+		LOG_ERR("TISCI Response in NACK\n");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int tisci_init(const struct device *dev)
+{
+	const struct tisci_config *config = dev->config;
+	struct tisci_data *data = dev->data;
+	int ret;
+
+	rx_message.buf = rx_message_buf;
+	rx_message.size = sizeof(rx_message_buf);
+
+	k_sem_init(&response_ready_sem, 0, 1);
+	data->host_id = config->host_id;
+	data->seq = 0x0;
+
+	k_sem_init(&data->data_sem, 1, 1);
+
+	data->desc.max_msg_size = MAILBOX_MBOX_SIZE;
+	/* Conservative timeout */
+	data->desc.max_rx_timeout_ms = 10000;
+
+	ret = mbox_register_callback_dt(&config->mbox_rx, callback, &rx_message);
+
+	if (ret < 0) {
+		LOG_ERR("Could not register callback (%d)\n", ret);
+		return ret;
+	}
+
+	ret = mbox_set_enabled_dt(&config->mbox_rx, true);
+	if (ret < 0) {
+		LOG_ERR("Could not enable RX channel (%d)\n", ret);
+		return ret;
+	}
+	return 0;
+}
+
+/* Clock Management Functions */
+int ti_sci_cmd_get_clock_state(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			       uint8_t *programmed_state, uint8_t *current_state)
+{
+	struct ti_sci_msg_resp_get_clock_state resp;
+	struct ti_sci_msg_req_get_clock_state req;
+	struct ti_sci_xfer *xfer;
+	int ret;
+
+	if (!dev) {
```

Comment:
You don't do anything with `dev` in this function, it is only used by `ti_sci_setup_one_xfer()`, so if you do this `dev` check inside the setup function, you can remove this check from a whole bunch of spots.